app/scripts/modules/core/src/securityGroup/securityGroupReader.service.ts

```diff
@@ -296,18 +302,37 @@ export class SecurityGroupReader {
   ) {}
 
   public getAllSecurityGroups(): PromiseLike<ISecurityGroupsByAccountSourceData> {
+    if (this.isFetchingGroups) {
+      return new Promise((resolve, reject) => {
+        this.awaitingGroupsResponseList.push([resolve, reject]);
+      });
+    }
     const cache = InfrastructureCaches.get('securityGroups');
     const cached = cache ? cache.get('allGroups') : null;
     if (cached) {
       return this.$q.resolve(this.decompress(cloneDeep(cached)));
     }
+    this.isFetchingGroups = true;
     return REST('/securityGroups')
       .get()
       .then((groupsByAccount: ISecurityGroupsByAccountSourceData) => {
         if (cache) {
           cache.put('allGroups', this.compress(groupsByAccount));
         }
+        for (const [resolve, _] of this.awaitingGroupsResponseList) {
+          resolve(groupsByAccount);
+        }
         return groupsByAccount;
+      })
+      .catch((e) => {
+        for (const [_, reject] of this.awaitingGroupsResponseList) {
+          reject(e);
+        }
+        throw new Error(e);
+      })
+      .finally(() => {
+        this.awaitingGroupsResponseList = [];
+        this.isFetchingGroups = false;
       });
   }
```

Comment:
I'll propose an alternate solution with less state management:

```suggestion
  private getAllSecurityGroupsPromise: PromiseLike<ISecurityGroupsByAccountSourceData>;

  public getAllSecurityGroups(): PromiseLike<ISecurityGroupsByAccountSourceData> {
    const cache = InfrastructureCaches.get('securityGroups');
    const cached = cache ? cache.get('allGroups') : null;
    if (cached) {
      return this.$q.resolve(this.decompress(cloneDeep(cached)));
    } else if (this.getAllSecurityGroupsPromise) {
      return this.getAllSecurityGroupsPromise;
    }

    this.getAllSecurityGroupsPromise = REST('/securityGroups')
      .get()
      .then((groupsByAccount: ISecurityGroupsByAccountSourceData) => {
        if (cache) {
          cache.put('allGroups', this.compress(groupsByAccount));
        }
        return groupsByAccount;
      })
      .finally(() => {
        delete this.getAllSecurityGroupsPromise;
      });

    return this.getAllSecurityGroupsPromise;
  }
```

Reply:
That's nice! Thanks